### PR TITLE
Feat/upgradability

### PR DESCRIPTION
**Pull request by auditsbydanny**

### CODE REVIEW

File: contracts/src/Dependencies/AddRemoveManagers.sol [hunk 1]
```diff
@@ -2,12 +2,14 @@
...
 
+import "openzeppelin-contracts/contracts/proxy/utils/Initializable.sol";
+
 import "../Interfaces/IAddRemoveManagers.sol";
 import "../Interfaces/IAddressesRegistry.sol";
 import "../Interfaces/ITroveNFT.sol";
 
-contract AddRemoveManagers is IAddRemoveManagers {
-    ITroveNFT internal immutable troveNFT;
+abstract contract AddRemoveManagers is Initializable, IAddRemoveManagers {
```
Comment (0x73696d616f) [inline contracts/src/Dependencies/AddRemoveManagers.sol]: No need to make this abstracts

File: contracts/src/Dependencies/LiquityBase.sol [hunk 2]
```diff
@@ -2,37 +2,39 @@
...
+import "../Libraries/LiquityBaseInit.sol";
 
 // import "forge-std/console2.sol";
 
 /*
 * Base contract for TroveManager, BorrowerOperations and StabilityPool. Contains global system constants and
 * common functions.
 */
-contract LiquityBase is ILiquityBase {
+abstract contract LiquityBase is Initializable, ILiquityBase {
```
Comment (0x73696d616f) [inline contracts/src/Dependencies/LiquityBase.sol]: No need abstract

File: contracts/src/Libraries/BorrowerOperationsInit.sol [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+    event SortedTrovesAddressChanged(address _sortedTrovesAddress);
+    event BoldTokenAddressChanged(address _boldTokenAddress);
+
+    function initialize(IAddressesRegistry _addressesRegistry) external {
+        // Initialize LiquityBase
+        LiquityBaseInit.initialize(_addressesRegistry);
+
+        // Initialize AddRemoveManagers
+        {
+            address troveNFT = address(_addressesRegistry.troveNFT());
```
Comment (0x73696d616f) [inline contracts/src/Libraries/BorrowerOperationsInit.sol:46]: Missing event, let's place this one into a lib just like LiquityBaseInit to remove duplication

File: contracts/src/Libraries/BorrowerOperationsInit.sol [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+            assembly {
+                sstore(TROVE_NFT_SLOT, troveNFT)
+            }
+        }
+
+        // Initialize BorrowerOperations
+        {
```
Comment (0x73696d616f) [inline contracts/src/Libraries/BorrowerOperationsInit.sol:54]: The order here is different than the original one in the constructor, let's use the same order of vars to make it easier to debug

File: contracts/src/Libraries/LiquityBaseInit.sol [hunk 5]
```diff
@@ -0,0 +1,38 @@
...
+    event PriceFeedAddressChanged(address _newPriceFeedAddress);
+
+    function initialize(IAddressesRegistry _addressesRegistry) external {
+        address activePool = address(_addressesRegistry.activePool());
+        address defaultPool = address(_addressesRegistry.defaultPool());
+        address priceFeed = address(_addressesRegistry.priceFeed());
+
+        assembly {
+            let ptr := sload(ACTIVE_POOL_SLOT)
+            let mask := not(shl(mul(ACTIVE_POOL_OFFSET, 8), sub(shl(20, 1), 1)))
```
Comment (0x73696d616f) [inline contracts/src/Libraries/LiquityBaseInit.sol:31]: add 1 comment saying storage variables are written from right to left
ex: `uint8 public var = 1;` becomes 0x00...01

File: contracts/src/Zappers/GasCompZapper.sol [hunk 6]
```diff
@@ -12,16 +13,26 @@ import "../Dependencies/Constants.sol";
 
 // import "forge-std/console2.sol";
 
-contract GasCompZapper is AddRemoveManagers {
```
Comment (0x73696d616f) [inline contracts/src/Zappers/GasCompZapper.sol:15]: missing gap

File: contracts/src/scripts/DeployLiquity2.s.sol [hunk 7]
```diff
@@ -296,7 +328,14 @@ contract DeployLiquity2Script is Script, StdCheats, MetadataDeployment {
         return abi.encodePacked(_creationCode, abi.encode(_addressesRegistry));
     }
 
-    function _deployAndConnectContracts(TroveManagerParams[] memory troveManagerParamsArray, IWETH _WETH)
+    function getBytecodeProxy(address _impl, address _admin) public pure returns (bytes memory) {
+        return abi.encodePacked(type(TransparentUpgradeableProxy).creationCode, abi.encode(_impl, _admin, ""));
```
Comment (0x73696d616f) [inline contracts/src/scripts/DeployLiquity2.s.sol]: "" should be replaced by initialization data

File: contracts/src/scripts/DeployLiquity2.s.sol [hunk 8]
```diff
@@ -335,8 +374,17 @@ contract DeployLiquity2Script is Script, StdCheats, MetadataDeployment {
         }
 
         r.collateralRegistry = new CollateralRegistry(r.boldToken, vars.collaterals, vars.troveManagers);
-        r.hintHelpers = new HintHelpers(r.collateralRegistry);
-        r.multiTroveGetter = new MultiTroveGetter(r.collateralRegistry);
+
+        r.hintHelpers = HintHelpers(address(new TransparentUpgradeableProxy(address(hintHelpersImpl), address(proxyAdmin), "")));
```
Comment (0x73696d616f) [inline contracts/src/scripts/DeployLiquity2.s.sol]: let's initialize in the constructor with the initialization data to make it atomic

File: contracts/src/scripts/DeployLiquity2.s.sol [hunk 9]
```diff
@@ -335,8 +374,17 @@ contract DeployLiquity2Script is Script, StdCheats, MetadataDeployment {
...
+        r.hintHelpers.initialize(r.collateralRegistry);
+
+
+        r.multiTroveGetter = MultiTroveGetter(address(new TransparentUpgradeableProxy(
```
Comment (0x73696d616f) [inline contracts/src/scripts/DeployLiquity2.s.sol]: let's create a helper function to deploy the proxy with the implementation and the initialize data
will help reduce duplication a bit

File: contracts/src/scripts/DeployLiquity2.s.sol [hunk 10]
```diff
@@ -385,6 +440,7 @@ contract DeployLiquity2Script is Script, StdCheats, MetadataDeployment {
         IHintHelpers _hintHelpers,
         IMultiTroveGetter _multiTroveGetter
     ) internal returns (LiquityContractsTestnet memory contracts) {
+        SALT = keccak256(abi.encodePacked(block.timestamp, address(_addressesRegistry)));
```
Comment (0x73696d616f) [inline contracts/src/scripts/DeployLiquity2.s.sol]: SALT should be the original one, will work if we set the initialize arguments in the proxy

File: contracts/src/test/deployment.t.sol [hunk 11]
```diff
@@ -187,4 +188,47 @@ contract Deployment is DevTestSetup {
         assertEq(defaultPoolAddress, recordedDefaultPoolAddress);
     }
     */
+
+
+   function testBorrowerOperationsInitialized() public view {
```
Comment (0x73696d616f) [inline contracts/src/test/deployment.t.sol:193]: let's test the other upgradeable contracts too

**Review by 0x73696d616f — CHANGES_REQUESTED** [review]
Looking good, nice work on the assembly.
Some additional stuff
1. TroveNft and metadataNFT should be upgradeable
2. Add to the readme that we remove immutable vars to not assign some variables in the constructor and others in the initialize()
3. Add to the readme that forge inspect does not work for some reason and needs to be brute forced. But let's keep the contracts as is, forge inspect should only be called now and if we change the code in the future.
4. Let's add some tests to test upgrade with reinitializer() for each contract and change ownership in the ones with ownership

File: contracts/src/Dependencies/AddRemoveManagers.sol (hunk 1 above)
Comment (auditsbydanny) [reply contracts/src/Dependencies/AddRemoveManagers.sol]: Fixed in `a177249aa0db6de2a4a68bcd7ddd85f59b453da9`

File: contracts/src/Dependencies/LiquityBase.sol (hunk 2 above)
Comment (auditsbydanny) [reply contracts/src/Dependencies/LiquityBase.sol]: Fixed in `a177249aa0db6de2a4a68bcd7ddd85f59b453da9`

File: contracts/src/Libraries/BorrowerOperationsInit.sol (hunk 3 above)
Comment (auditsbydanny) [reply contracts/src/Libraries/BorrowerOperationsInit.sol:46]: Fixed in `6bddd9edaefcde040a584f9b4f298f3e2b95ea52`

File: contracts/src/Libraries/BorrowerOperationsInit.sol (hunk 4 above)
Comment (auditsbydanny) [reply contracts/src/Libraries/BorrowerOperationsInit.sol:54]: Fixed in `6bddd9edaefcde040a584f9b4f298f3e2b95ea52`

File: contracts/src/Libraries/LiquityBaseInit.sol (hunk 5 above)
Comment (auditsbydanny) [reply contracts/src/Libraries/LiquityBaseInit.sol:31]: Fixed in `6bddd9edaefcde040a584f9b4f298f3e2b95ea52`

File: contracts/src/Zappers/GasCompZapper.sol (hunk 6 above)
Comment (auditsbydanny) [reply contracts/src/Zappers/GasCompZapper.sol:15]: Fixed in `0302acdfe06290770603f0a13a394ee9497f81fa`

File: contracts/src/scripts/DeployLiquity2.s.sol (hunk 7 above)
Comment (auditsbydanny) [reply contracts/src/scripts/DeployLiquity2.s.sol]: Fixed in `d23e391f3697654901994020bffff25fd197614e`

File: contracts/src/scripts/DeployLiquity2.s.sol (hunk 8 above)
Comment (auditsbydanny) [reply contracts/src/scripts/DeployLiquity2.s.sol]: Fixed in `d23e391f3697654901994020bffff25fd197614e`

File: contracts/src/scripts/DeployLiquity2.s.sol (hunk 9 above)
Comment (auditsbydanny) [reply contracts/src/scripts/DeployLiquity2.s.sol]: Fixed in `d23e391f3697654901994020bffff25fd197614e`

File: contracts/src/scripts/DeployLiquity2.s.sol (hunk 10 above)
Comment (auditsbydanny) [reply contracts/src/scripts/DeployLiquity2.s.sol]: Fixed in `d23e391f3697654901994020bffff25fd197614e`

**Pull request merged**